cmd/admin-handlers.go

```diff
@@ -1303,6 +1304,155 @@ func (a adminAPIHandlers) KMSKeyStatusHandler(w http.ResponseWriter, r *http.Req
 }
 
+// KMSKeyRotateHandler - POST /minio/admin/v2/kms/key/rotate
+func (a adminAPIHandlers) KMSKeyRotateHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := newContext(r, w, "KMSKeyRotateHandler")
+
+	objectAPI, _ := validateAdminReq(ctx, w, r, iampolicy.KMSKeyRotateAdminAction)
+	if objectAPI == nil {
+		return
+	}
+
+	if GlobalKMS == nil {
+		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrKMSNotConfigured), r.URL)
+		return
+	}
+
+	var req madmin.KMSRotateKeyRequest
+	if err := json.NewDecoder(io.LimitReader(r.Body, 1<<20)).Decode(&req); err != nil {
+		writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+		return
+	}
+
+	// Delete rotation entries where the old key ID
+	// is equal to the new key ID. It's pointless to
+	// rotate when the old and the new key are the same.
+	for k, v := range req.KeyMapping {
+		if k == v {
+			delete(req.KeyMapping, k)
+		}
+	}
+
+	// Verify that the key mapping is well-formed. In particular,
+	// that a old key ID does not appear as new key ID as well.
+	// This is not allowed since it would cause unpredictable
+	// behavior.
+	for oldKey := range req.KeyMapping {
+		for _, newKey := range req.KeyMapping {
+			if oldKey == newKey {
+				err := AdminError{
+					StatusCode: http.StatusBadRequest,
+					Code:       http.StatusText(http.StatusBadRequest),
+					Message:    fmt.Sprintf("invalid key mapping: cannot rotate from and to: %s", oldKey),
+				}
+				writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+				return
+			}
+		}
+	}
+
+	// If there is no key mapping for rotation return early.
+	if len(req.KeyMapping) == 0 {
+		writeSuccessResponseHeadersOnly(w)
+		return
+	}
+
+	// If the bucket is empty and rotation is not recursive there are no
+	// objects - only buckets. So return early.
+	if req.Bucket == "" && !req.Recursive {
+		writeSuccessResponseHeadersOnly(w)
+		return
+	}
+
+	// Either check whether the requested bucket exists or
+	// if no bucket is specified list all buckets.
+	var bucketNames []string
+	if req.Bucket != "" {
+		if _, err := objectAPI.GetBucketInfo(ctx, req.Bucket); err != nil {
+			writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+			return
+		}
+		bucketNames = []string{req.Bucket}
+	} else {
+		bucketInfo, err := objectAPI.ListBuckets(ctx)
+		if err != nil {
+			writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+			return
+		}
+		bucketNames = make([]string, len(bucketInfo))
+		for i, b := range bucketInfo {
+			bucketNames[i] = b.Name
+		}
+	}
+
+	// For each bucket list all objects for the requested prefix.
+	// If the object is encrypted with the KMS perform a key
+	// rotation if there is a mapping for the current key ID.
+	for _, bucket := range bucketNames {
+		var err error
+		var objectListing ListObjectsV2Info
+		for {
+			objectListing, err = objectAPI.ListObjectsV2(ctx, bucket, req.Prefix, objectListing.NextContinuationToken, "", 1000, false, objectListing.ContinuationToken)
```

Comment:
The timeout will happen simply because batched listing might get slower even for the first batch. I would suggest perhaps implement functionality similar to HealObjects() or have a Walk kind of API at the object layer which helps you get the list as soon as possible, as there is no real use here for batched operation. The batched operation has more state and is not quite useful in your scenario.

It would be very simple like this

```
doneCh := sendWhiteSpaceToHTTPResponse(w)
walkCh, err := objAPI.Walk(...)
for {
select objInfo := <-walkCh:
 ... all my stuff ... 
select <-doneCh:
        return
}
```

Reply:
Are you suggesting to implement a generic `Walk` function in the object layer? This sounds good to me. However, should this be designed to be a generic building block for healing, key rotation, key listing, etc.? `Walk` seems something we need across API functions...

Reply:
You don't need to worry about building it to be generic on the get go. HealObjects already uses Walk without batched listing. 

You need to implement this for yourself, just look at HealObjects on how it can be implemented. All this API does is returns object info over a channel and you just rotate those objects one at a time.

Reply:
Yeah, that I got. My point being that a `Walk` functionality seems to be common pattern in the server code. Therefore, asking whether it's worth unifying to a common API. (Not in this PR but separate).